### PR TITLE
added some metabolic functions

### PR DESCRIPTION
**Pull request by jackr555**
# Description

I have added equations 8 to 13. However I am unsure how to write the sum term from equation 1.

Fixes # (issue)

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [ ] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [ ] All tests pass: `$ poetry run pytest`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 1]
```diff
@@ -112,6 +112,108 @@ def da(
     return uda
 
 
+# metabolic functions
+
+
+def calculate_kappa(
+    Gatp: float = 75,
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: In the paper this is 75 kJ mol^-1, so to be consistent with G0 (which is in J mol^-1) this should be 75000

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 2]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    G0: float = 1.5e5,
+    reaction_energy: float = 1 / 3,
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: I wouldn't give this variable a default value, instead leave it as a variable and then call `calculate_kappa` with a specific value for it i.e. `calculate_kappa(3.0)`

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 3]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    return u
+
+
+def calculate_theta(s: float, w: float) -> float:
+    """A thermodynamic factor that stops a reaction at equilibrium.
+
+    Args:
+     s: substrate concentration
+     w: waste product concentration
+     kappa: species equilibrium constant.
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: kappa isn't an argument

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 4]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    """
+    kappa = calculate_kappa()
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: `calculate_kappa` should be called with the value for `reaction_energy` provided

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 5]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    u = (w / s) / kappa
+    return u
+
+
+def calculate_E(v: float, R: float, Q: float = 0.45, m: float = 1e8) -> float:
+    """Calculates enzyme copy number."""
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: This is good, but would be good to have the arguments described

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 6]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    R: float,
+    s: float,
+    w: float,
+    ka: float = 10,
+    ks: float = 1.375e-3,
+    ra: float = 10,
+) -> float:
+    """Calculates reaction rate."""
+    E = calculate_E(1, R)
+    calculate_kappa(Gatp)
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: You don't need to call this function as it's called within `calculate_theta`

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 7]
```diff
@@ -112,6 +112,108 @@ def da(
...
+    theta = calculate_theta(s, w)
+    u = (ka * E * s * (1 - theta)) / (ks + s * (1 + ra * theta))
+    return u
+
+
+def j(
+    Gatp: float, R: float, s: float, w: float, reaction_energy: float = 1 / 3
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: I don't think you should give a default value for `reaction_energy`

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 8]
```diff
@@ -112,6 +112,108 @@ def da(
...
+
+def calculate_consumed(Gatp: float, R: float, s: float, w: float) -> float:
+    """Calculates rate a metabolite is consumed per cell."""
+    u = q(Gatp, R, s, w)
+    return u
+
+
+# below is the code i could not get working
+"""
+def calculate_sum_of_species(num_species: float = 2, N: float)-> float:
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: Okay so what you really need here is an "if else" statement.

```python
if substrate_number == 1:
      consumed = calculate_consumed(R, s=c[0], w=[0])
      u = -consumed*N
else:
     produced = calculate_produced(R, s=c[0], w=[0])
     u = produced*N
return u
```

Comment (jacobcook1995) [reply microbial_thermodynamics/cell_growth_ODE.py]: That should work because what you are supplying the function is not a float but a vector of floats so `consumed` is a vector which then gets multiplied by a vector

**Review by jacobcook1995 — CHANGES_REQUESTED** [review]
The code quality is really good, well done! 

I've made some comments on things to improve and also some suggestions of how to get `calculate_sum_of_species` to work. I think it might be worth meeting to discuss that a bit though

File: microbial_thermodynamics/scipy_demo.py [hunk 9]
```diff
@@ -50,10 +49,20 @@ def full_equation_set(
...
+        w=y[7 * number_of_species : 8 * number_of_species],
+        reaction_energy=reaction_energies,
+        v=y[5 * number_of_species : 6 * number_of_species],
+    )
+    change_in_c = dc(
+        c=y[3 * number_of_species : 4 * number_of_species],
+        reaction_energy=reaction_energies,
+        N=y[0:number_of_species],
+        R=y[number_of_species : 2 * number_of_species],
+        v=y[5 * number_of_species : 6 * number_of_species],
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: `number_of_species = 2` and y0 has 8 elements so these numbers go too high i.e. `y[10:12]`

File: microbial_thermodynamics/scipy_demo.py [hunk 10]
```diff
@@ -50,10 +49,20 @@ def full_equation_set(
         t,
         a=y[2 * number_of_species : 3 * number_of_species],
         R=y[number_of_species : 2 * number_of_species],
+        s=y[6 * number_of_species : 7 * number_of_species],
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: You should be selecting a vector here just a single value, so `s=y[3*number_of_species]` and `w=y[3*number_of_species + 1]`

File: microbial_thermodynamics/scipy_demo.py [hunk 11]
```diff
@@ -50,10 +49,20 @@ def full_equation_set(
...
+        w=y[7 * number_of_species : 8 * number_of_species],
+        reaction_energy=reaction_energies,
+        v=y[5 * number_of_species : 6 * number_of_species],
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: `v` isn't containing the the `y` vector (as it doesn't get updated by the integration process), for now I would just use `v=[1.0, 1.0]` as each species only has one reaction

**Review by jacobcook1995 — CHANGES_REQUESTED** [review]
Here are some suggestions to fix how v, s and w are used. Hopefully the code will run once these are sorted

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 12]
```diff
@@ -94,24 +94,156 @@ def dr(tr: float, a: float, R: float) -> float:
...
+    """Calculates the kappa constant.
+
+    Args:
+     reaction_energy: the ATP generated for each species for each reaction alpha
+     Gatp: ATP free energy
+     G0: the standard Gibbs free-energy change when one mole of reaction alpha occurs
+     R: the gas constant
+     T: temperature
+    """
+    u = np.exp((((-G0) - reaction_energy) * Gatp) / R * T)
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: Okay this equation is slightly wrong, as you multiple both `G0` and `reaction_energy` (eta in the paper) by `Gatp`. In fact it should only be `reaction_energy` multiplied by `Gatp`. Multiplying `G0` by `Gatp` results in a very large negative number which means that u goes to 0.

I'm pretty sure this is what the correct expression should be (though check it against the paper).
```python
u = np.exp(((-G0) - reaction_energy * Gatp) / R * T)
```

File: microbial_thermodynamics/scipy_demo.py [hunk 13]
```diff
@@ -50,10 +49,20 @@ def full_equation_set(
         t,
         a=y[2 * number_of_species : 3 * number_of_species],
         R=y[number_of_species : 2 * number_of_species],
+        s=y[3 * number_of_species],
+        w=y[3 * number_of_species + 1],
+        reaction_energy=reaction_energies,
+        v=[1.0, 2.0],
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: If you change this (and the one below) to `v=np.array([1.0, 2.0])` does that work

Comment (jacobcook1995) [reply microbial_thermodynamics/scipy_demo.py]: I think it might work with a `numpy` array, but doesn't work with lists of floats

Comment (jackr555) [reply microbial_thermodynamics/scipy_demo.py]: No I still get the message TypeError: unsupported operand type(s) for -: 'float' and 'list'

Comment (jacobcook1995) [reply microbial_thermodynamics/scipy_demo.py]: Does it work when you change the type hint for v to `NDArray[np.float32]`?

Comment (jackr555) [reply microbial_thermodynamics/scipy_demo.py]: I currently have the type hint for v and Reaction_energy set to NDArray[np.float32]. I have pushed what it currently looks like.

File: microbial_thermodynamics/scipy_demo.py [hunk 14]
```diff
@@ -75,16 +84,15 @@ def integrate() -> NDArray[np.float32]:
     N0 = [23.0, 2.3]
     r0 = [0.3, 0.3]
     a0 = [1e6, 1e6]
+    c0 = [0, 0]
 
     # Construct vector of initial values y0
-    y0 = np.concatenate((N0, r0, a0))
+    y0 = np.concatenate((N0, r0, a0, c0))
+    reaction_energies = [1.0, 2.0]
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: This one should be `np.array([1.0,2.0])`

File: microbial_thermodynamics/scipy_demo.py [hunk 15]
```diff
@@ -50,10 +49,20 @@ def full_equation_set(
         t,
         a=y[2 * number_of_species : 3 * number_of_species],
         R=y[number_of_species : 2 * number_of_species],
+        s=y[3 * number_of_species],
+        w=y[3 * number_of_species + 1],
+        reaction_energy=reaction_energies,
+        v=np.array([1.0, 2.0]),
```
Comment (jacobcook1995) [inline microbial_thermodynamics/scipy_demo.py]: This and the one below should be `v=np.array([1.0, 1.0])` (this is unrelated to the current problem, it's just a mistake that I noticed)

**Comment by jackr555** [comment]
now there appears to be a problem with the dc function. ValueError: setting an array element with a sequence.

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 16]
```diff
@@ -94,22 +94,171 @@ def dr(tr: float, a: float, R: float) -> float:
...
+        if ind == 0:
+            c_changes[ind] = (
+                k
+                - p * c[ind]
+                - N[ind] * q(R[ind], c[0], c[1], reaction_energy[ind], v[ind])
+            )
+        else:
+            c_changes[ind] = -p * c[ind] + N[ind] * q(
+                R[ind], c[0], c[1], reaction_energy[ind], v[ind]
+            )
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py:283]: Okay so the above should become
```python
for metabolite_num, _ in enumerate(c):
    for species_num, _ in enumerate(N):
        if metabolite_num == 0:
            c_changes[metabolite_num] += (
                k
                - p * c[metabolite_num]
                - N[species_num] * q(R[species_num], c[0], c[1], reaction_energy[species_num], v[species_num])
            )
        else:
            c_changes[metabolite_num] += -p * c[metabolite_num] + N[species_num] * q(
                R[species_num], c[0], c[1], reaction_energy[species_num], v[species_num]
            )
```

Comment (jacobcook1995) [reply microbial_thermodynamics/cell_growth_ODE.py:283]: Basically need to loop over all metabolites, but then for each metabolite calculate the contribution for each species (This is done by looping over each species)

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 17]
```diff
@@ -94,22 +94,196 @@ def dr(tr: float, a: float, R: float) -> float:
...
+
+def dc(
+    c: NDArray[np.float32],
+    reaction_energy: NDArray[np.float32],
+    N: NDArray[np.float32],
+    R: NDArray[np.float32],
+    v: NDArray[np.float32],
+    k: float = 3.3e-7,
+    p: float = 6e-5,
+) -> float:
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: So a new input should be added here 
```python
avogadros_number: float = 6.022e23
```

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 18]
```diff
@@ -94,22 +94,196 @@ def dr(tr: float, a: float, R: float) -> float:
...
+     k: substrate supply rate
+     p: metabolite dilution rate
+    """
+    c_changes = np.zeros(len(c))
+    for ind, _ in enumerate(c):
+        if ind == 0:
+            c_changes[ind] = (
+                k
+                - p * c[ind]
+                - N[ind] * q(R[ind], c[0], c[1], reaction_energy[ind], v[ind])
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py]: This should become 
```python
- N[ind] * q(R[ind], c[0], c[1], reaction_energy[ind], v[ind])/avogadros_number
```

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 19]
```diff
@@ -94,22 +94,196 @@ def dr(tr: float, a: float, R: float) -> float:
...
+                        k
+                        - p * c[metabolite_num]
+                        - N[species_num]
+                        * q(
+                            R[species_num],
+                            c[0],
+                            c[1],
+                            reaction_energy[species_num],
+                            v[species_num],
+                        )
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py:298]: add `/ avogadros_number` after the `q(...)` to change the unit

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 20]
```diff
@@ -94,22 +94,196 @@ def dr(tr: float, a: float, R: float) -> float:
...
+                else:
+                    c_changes[metabolite_num] += -p * c[metabolite_num] + N[
+                        species_num
+                    ] * q(
+                        R[species_num],
+                        c[0],
+                        c[1],
+                        reaction_energy[species_num],
+                        v[species_num],
+                    )
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py:313]: and again add / avogadros_number after the q(...) to change the unit

**Comment by jacobcook1995** [comment]
Okay so to plot values generated by functions (e.g. lambda) using the integration output you need to do the following:
```python
output = integrate()
plot_time = output.t
# Find a and R values for species 1 for the whole simulation
species_1_a = output.y[4]
species_1_R = output.y[2]
# Use the a and R values to calculate lambda from the function you already defined
species_1_lambda = calculate_lam(a=species_1_a, R=species_1_R)
# Then plot lambda against time
plt.plot(plot_time, species_1_lambda)
```

**Comment by jacobcook1995** [comment]
To loop over the `j` function you would do something like this:
```python
output = integrate()
plot_time = output.t
species_1_R = output.y[2]
metabolite_1 = output.y[6]
metabolite_2 = output.y[7]
# Create empty vector to store j value for each time point in plot_time
species_1_j = = np.zeros(len(plot_time))
for ind, _ in enumerate(plot_time):
    species_1_j[ind] = calculate_j(R=species_1_R[ind], s= metabolite_1[ind], w= metabolite_2[ind], v = 1.0)
# Then plot species_1_j
```

File: microbial_thermodynamics/cell_growth_ODE.py [hunk 21]
```diff
@@ -94,22 +94,203 @@ def dr(tr: float, a: float, R: float) -> float:
     return udr
```
Comment (jacobcook1995) [inline microbial_thermodynamics/cell_growth_ODE.py:110]: Okay in the above the expression is:
```python
udr = (1 / time_scale) * (r_star * a - R)
```
but from the paper it should be:
```python
udr = (1 / time_scale) * (r_star - R)
```

Comment (jacobcook1995) [reply microbial_thermodynamics/cell_growth_ODE.py:110]: So that when `r_star = R`, `udr=0` and r stops changing.

**Review by jacobcook1995 — APPROVED** [review]
